ml-agents/mlagents/trainers/env_manager.py

```diff
@@ -32,19 +29,28 @@ def empty(worker_id: int) -> "EnvironmentStep":
 
 class EnvManager(ABC):
     def __init__(self):
-        self.policies: Dict[AgentGroup, Policy] = {}
+        self.policies: Dict[AgentGroup, TFPolicy] = {}
+        self.agent_managers: Dict[AgentGroup, AgentManager] = {}
 
-    def set_policy(self, brain_name: AgentGroup, policy: Policy) -> None:
+    def set_policy(self, brain_name: AgentGroup, policy: TFPolicy) -> None:
         self.policies[brain_name] = policy
+        if brain_name in self.agent_managers:
+            self.agent_managers[brain_name].policy = policy
+
+    def set_agent_manager(self, brain_name: AgentGroup, manager: AgentManager) -> None:
+        self.agent_managers[brain_name] = manager
 
     @abstractmethod
-    def step(self) -> List[EnvironmentStep]:
+    def _step(self) -> List[EnvironmentStep]:
         pass
 
     @abstractmethod
-    def reset(self, config: Dict = None) -> List[EnvironmentStep]:
+    def _reset_env(self, config: Dict = None) -> List[EnvironmentStep]:
```

Comment:
_reset_env returns a List[EnvironmentStep], but should it only contain one (or none at all ?
advance will return the num_step_infos added but reset will not. Why is that different if both can process a list of step infos?

Reply:
I believe _reset_env could return multiple EnvironmentSteps if there's multiple workers; they're the first obs after the reset. We don't need the number of them since that number is used to step the global counter, and that counter should start at 0 on reset. But I can return their number anyways if it's cleaner.

Reply:
The counter should reset at 0 but if you received a bunch of environment steps when calling reset, that number should be increased anyway no? That is because reset does not only reset. It resets and steps the environments a bunch.

Reply:
The steps aren't steps though right - they're the 0th timestep (the initial state of the environment before taking any steps) from the RL perspective